[SYCL][RFC] SYCL runtime upstreaming RFC (#11540)
intel/llvm: sycl/doc/SYCLRuntimeRFC.md

various backends in a similar fashion to libomptarget.

This RFC contains a brief overview of the core functionality of the SYCL runtime
and its major components. The main RFC for SYCL implementation upstreaming can
be found here:
https://discourse.llvm.org/t/rfc-add-full-support-for-the-sycl-programming-model/74080

## Host and device code integration

Consider this SYCL code:

```
#include <sycl/sycl.hpp>
int main() {
  sycl::queue q;
  sycl::range<3> range{8, 8, 8};
  sycl::buffer<double, 3> b{range};
  int i = 12;
  q.submit([&](sycl::handler &cgh) {
    sycl::accessor acc{b, cgh};
    cgh.parallel_for(range, [=](sycl::id<3> idx) {
      acc[idx] += i;
    });
  });
}
```

The lambda passed to the `sycl::handler::parallel_for` function represents a
device kernel. When a SYCL application is started or a dynamic SYCL library is
loaded, all device images contained in the multi-targeted binary are registered
in the program manager, which is one of the larger components of the SYCL
runtime, using SYCL's `__tgt_register_lib`
(https://clang.llvm.org/docs/OffloadingDesign.html#global-constructor-and-destructor)

> counterpart, `__sycl_register_lib`. The device image wrappers generated by SYCL

Reviewer: This reads a little odd to me. I first read it as stating that `__tgt_register_lib` is part of SYCL because of the distance separation from "counterpart" introduced by the link. Also, `__sycl_register_lib` isn't mentioned in the SYCL specification, so I'm not sure what this is trying to say.

Author: It's just supposed to say that internally we use the same mechanism for device image registration as OpenMP with its '__tgt_register_lib`. Changed the wording to make the link placement less awkward.

build options. The integration header, which is generated during device compilation and
included during host compilation (see
https://discourse.llvm.org/t/rfc-sycl-host-compiler-integration-header-and-footer/74081),
helps the runtime map SYCL kernel invocations to their corresponding symbols in
the device code. The program manager then uses information from device image
wrappers to map the kernel name to a set of device images, chooses an image
compatible with the requested device, passes it to the backend to be
just-in-time compiled if needed, and caches the result for reuse.

The integration header also provides information about kernel arguments.
It assumes that the lambda layout is the same for both device and host
compilation, and the runtime uses the information about argument size
and offsets to extract their values from the lambda object passed to
`sycl::handler::parallel_for` and sets them when enqueuing the kernel. The
integration header also includes information about argument types, which
is needed to handle special cases like `sycl::accessor` objects.

For more details about device code compilation, see

> https://discourse.llvm.org/t/rfc-sycl-kernel-lowering/74082
> https://discourse.llvm.org/t/rfc-offloading-design-for-sycl-offload-kind-and-spir-targets/74088

Reviewer: ```suggestion
https://discourse.llvm.org/t/rfc-sycl-kernel-lowering/74082
and
https://discourse.llvm.org/t/rfc-offloading-design-for-sycl-offload-kind-and-spir-targets/74088
```
The fact that there are two links there is hard to see; add an "and".

> scourse topic details

Reviewer 2: Nit
```suggestion
Discourse topic details
```

- Category: [Runtimes](https://discourse.llvm.org/c/runtimes/64)
- Title: "RFC: SYCL runtime upstreaming"
- Tags: sycl

## Overview

> Our SYCL runtime provides an implementation of the SYCL 2020 API specification.

Reviewer 2: ```suggestion
SYCL runtime provides an implementation of the SYCL 2020 API specification.
```

It's responsible for managing resources for enqueuing tasks to the offload
device, tracking dependencies between them, and data movement between the host
and devices. The SYCL runtime is device-agnostic and uses Unified Runtime

> (https://github.com/oneapi-src/unified-runtime) as an external dependency that

Reviewer 2: ```suggestion
(https://github.com/oneapi-src/unified-runtime) as an external dependency. This Unified Runtime
```

serves as an interface layer between the SYCL runtime and device-specific
backends. Unified Runtime has several adapters that bind to various backends in
a similar fashion to libomptarget.

> This RFC contains a brief overview of the core functionality of the SYCL runtime

Reviewer 2: Is there any specific question/comment that we will need to ask the community here?

and its major components. The main RFC for SYCL implementation upstreaming can

> be found here: (TODO add link to the main RFC).

Author: Added.

> ## Host and device code integration

Reviewer 2: Nit: May be "SYCL Program Manager" will be a better heading here?

Author: This talks about more than just that, it just so happens that program manager plays a substantial part. I'd rather keep it as is.

Consider this SYCL code:

```
#include <sycl.hpp>
int main() {
  sycl::queue q;
  sycl::buffer<char> b{sycl::range{1024}};
  q.submit([&](sycl::handler &cgh) {
    sycl::accessor acc{b, cgh};
    int i = 13;
    struct S {
      char c;
      int i;
    } test_s;
    test_s.c = 14;
    cgh.single_task([=] {
      if (i == 13 && test_s.c == 14) {

>         acc[0] = 'a';

Author: That's fair, I took it from the integration header RFC, but it's probably way more appropriate there to illustrate how different captured types are handled.
Updated.

      }
    });
  });
}
```

The lambda passed to the `sycl::handler::single_task` function
represents a device kernel. When a SYCL application is started or a
dynamic SYCL library is loaded, all device images contained in the

> multi-targeted binary are registered in the program manager, one of the

Reviewer 2: ```suggestion
multi-targeted binary are registered in the program manager, which is one of the
```

> larger components of the SYCL runtime, using SYCL's `__tgt_register_lib`

Reviewer 2: What is __tgt_register_lib? Are readers of RFC going to be aware? May be a line about it? \

Thanks

Reviewer 3: It's LLVM's standard symbols described in https://clang.llvm.org/docs/OffloadingDesign.html, so it reasonable to expect RFC reviewers to be aware of it. It might be helpful to provide a link to the LLVM's documentation anyway.

Author: Added a link.

> counterpart, `__sycl_register_lib`. The integration header, which is

Reviewer 2: May be a link to the integration header RFC here?

Thanks

> generated during device compilation and pre-included during host

Reviewer 2: ```suggestion
generated during device compilation and included during host
```

compilation, helps the runtime map SYCL kernel invocations to their
corresponding symbols in the device code. The program manager uses the
information contained in device image wrappers to map the kernel name to
a set of device images, chooses an image compatible with the requested
device, passes it to the backend to be just-in-time compiled if needed,
and caches the result for reuse.

The integration header also provides information about kernel arguments.

> It assumes that the lambda layout is the same for both device and host

Reviewer 2: What's a 'lambda layout'?

Reviewer 3: Memory layout of the SYCL kernel object.

compilation, and the runtime uses the information about argument size
and offsets to extract their values from the lambda object passed to

> `sycl::handler::single_task` and set them when enqueuing the kernel. The

Reviewer 2: ```suggestion
`sycl::handler::single_task` and sets them when enqueuing the kernel. The
```

integration header also includes information about argument types, which
is needed to handle special cases like `sycl::accessor` objects. For

> more details about the integration header, see (TODO add link to the

Reviewer 2: This link will be more useful if provided at the beginning of the discussion.

header/footer RFC).

## Scheduler

> Another large runtime component is the scheduler. The scheduler

Reviewer 2: ```suggestion
Another large component in SYCL runtime is the scheduler. The scheduler
```

maintains a directed acyclic graph representation of the submitted

> command groups, managing dependencies and data movement required by the

Reviewer 2: This RFC seems to expect readers to have some idea about 'command groups', 'accessors', amongst other things. Do we want to add a link to SYCL spec or some other high level document?

Thanks

Author: Added a link to a relevant SYCL spec section

application code explicitly or with the use of accessors. Most of the

> actual synchronization is handled by the underlying device backend, with

Reviewer 2: synchronization between what?

the scheduler simply enqueuing tasks as they are submitted and passing a

> list of deduced dependencies. There are some exceptions to this: a SYCL

Reviewer 2: What is a 'deduced dependency'? May be a quick note can be added? Or we can just call this dependency.

> list of deduced dependencies. There are some exceptions to this: a SYCL

Reviewer 2: ```suggestion
list of deduced dependencies. There are some exceptions to this. A SYCL
```

> application may require access to some data on the host or submit an

Reviewer 2: ```suggestion
application may require access to some data on the host or may need to submit an
```

asynchronous host task. Such cases are handled by the scheduler directly

> by delaying the submission of device workloads until their host

Reviewer 2: To keep up with previous wording.
```suggestion
by delaying the submission of device tasks until their host
```

dependencies are satisfied.

## Device headers and libraries

> SYCL runtime headers provide various algorithms and built-in functions

Reviewer 2: ```suggestion
SYCL runtime headers and libraries provide various algorithms and built-in functions
```

Author: This was actually intended to refer to headers only. Changed "provide"->"declare" to make the intent clearer.

> to be used in device code. A large portion of these are simple enough to

Reviewer 2: ```suggestion
to be used in device code. A large portion of such algorithms and built-in functions are simple enough to
```

be implemented in the headers (for example, those directly lowered into
SPIR-V built-ins), but some have more complex implementations (for
example, fallback implementations for functionality that might not be

> supported natively). Those are available in the form of several

Reviewer 2: ```suggestion
supported natively). The more complex implementations are available in the form of several
```

> pre-built device libraries that are linked with the user device code

Reviewer 2: ```suggestion
pre-built device libraries that will later be linked with the user device code
```

Author: I prefer my wording here. This talks about the workflow in general and it's not clear what "later" is relative to in this sentence.

either during ahead-of-time compilation or at runtime by the program
manager, using the information about device library dependencies

> embedded in device image wrappers by SYCL compilation tools.

Reviewer 3: `libdevice` duplicates the functionality of libc and libm. I expect the AMD folks will ask us to use libc/libm instead of cotributing `libdevice` in its current form.

@jinge90, have you looked at libm/libc implementation? If yes, do you have a plan to integrate it into our implementation?

Reviewer 4: There was a presentation about using libm from llvm.  The results are mixed results when comparing the routines from libm  with vendor provided implementation.

Reviewer 3: During GPU offloading workshop we talked about having an option to customize libm implementation based on accuracy requirements. I expect this option to enable the use of SPIR-V math functions for libm implementation on Intel GPU.
AFAIK, @andykaylor is doing some work in this area.

Reviewer 5: Hi, @bader 
I don't have detail plan to integrate llvm libc/libm into our implementation for now but I agree that we should align with LLVM libc/libm for libdevice in sycl. I have looked at libm/libc before and most generic versions of functions can be directly used in sycl such as <string.h>, <math.h>.
For the first step, maybe we can start from the LLVM libc/libm building system to support to use sycl toolchain to build required functions.
Thanks very much.

Author: @bader @jinge90 Do you think it makes sense to mention that we're looking in this direction in this RFC then? If so, how would you have it worded?

Reviewer 4: @sergey-semenov not clear if Unified Runtime is sycl specific plugin or general UR which OpenMP is UR_plugin will  interface.  If it is  a plugin better to mention since we use similar  terms in OpenMP

Reviewer 3: I think so. It seems there is not much we can say at this point, except that "some device libraries functionality overlaps with libc/libm implementation for GPU. We are looking for ways to re-use LLVM code for SYCL and welcome any feedback".

This RFC contains a brief overview of the core functionality of the SYCL runtime
and its major components. The main RFC for SYCL implementation upstreaming can
be found here:
https://discourse.llvm.org/t/rfc-add-full-support-for-the-sycl-programming-model/74080

## Host and device code integration

Consider this SYCL code:

```
#include <sycl/sycl.hpp>
int main() {
  sycl::queue q;
  sycl::range<3> range{8, 8, 8};
  sycl::buffer<double, 3> b{range};
  int i = 12;
  q.submit([&](sycl::handler &cgh) {
    sycl::accessor acc{b, cgh};
    cgh.parallel_for(range, [=](sycl::id<3> idx) {
      acc[idx] += i;
    });
  });
}
```

The lambda passed to the `sycl::handler::parallel_for` function represents a
device kernel. When a SYCL application is started or a dynamic SYCL library is
loaded, all device images contained in the multi-targeted binary are registered
in the program manager, which is one of the larger components of the SYCL
runtime, using SYCL's `__tgt_register_lib`
(https://clang.llvm.org/docs/OffloadingDesign.html#global-constructor-and-destructor)
counterpart, `__sycl_register_lib`. The device image wrappers generated by SYCL

> compilation tools also include infromation like device kernel symbols and

Reviewer 2: ```suggestion
compilation tools also include information like device kernel symbols and
```

device-specific backends. Unified Runtime has several adapters that bind to
various backends in a similar fashion to libomptarget.

> The SYCL runtime uses semantic versioning (libsycl.so.major.minor.patch, see

Author: It's mostly to mention the somewhat unconventional major version postfix on Windows and the fact that the driver needs to be aware of the current library version for linking. Not terribly important, so I'm not opposed to leaving this out if others feel like mentioning this detail is unnecessary.
